File: src/main/java/za/co/absa/abris/utils/avro/kafka/write/KafkaAvroWriter.java
/*
 * Copyright 2018 ABSA Group Limited
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package za.co.absa.abris.utils.avro.kafka.write;

import java.util.List;
import java.util.Objects;
import java.util.Properties;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.TimeoutException;

import org.apache.kafka.clients.producer.KafkaProducer;
import org.apache.kafka.clients.producer.ProducerRecord;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import za.co.absa.abris.utils.avro.AvroPayloadConverter;

public class KafkaAvroWriter<T> {

	private final static Logger logger = LoggerFactory.getLogger(KafkaAvroWriter.class);
	
	private final static String PROP_BOOTSTRAP_SERVERS = "bootstrap.servers";
	private final static String PROP_METADATA_BROKER_LIST = "metadata.broker.list";
	private final static String PROP_CLIENT_ID = "client.id";
	private final static String PROP_ACKS = "acks";
	private final static String PROP_KEY_SERIALIZERS = "key.serializer";
	private final static String PROP_VALUE_SERIALIZER = "value.serializer";	

	private final KafkaProducer<String, byte[]> kafkaSender;
	private final AvroPayloadConverter avroConverter;

	public KafkaAvroWriter(Properties connectionProps) {
		if (!validate(connectionProps)) {
			throw new IllegalArgumentException("Missing Kafka connection parameters.");
		}		
		
		this.kafkaSender = new KafkaProducer<String, byte[]>(connectionProps);	
		this.avroConverter = new AvroPayloadConverter();
	}

	private final boolean validate(Properties properties) {
		return contains(properties, PROP_BOOTSTRAP_SERVERS) && 
				contains(properties, PROP_METADATA_BROKER_LIST) &&
				contains(properties, PROP_CLIENT_ID) &&
				contains(properties, PROP_ACKS) &&
				contains(properties, PROP_KEY_SERIALIZERS) &&
				contains(properties, PROP_VALUE_SERIALIZER);
	}

	private final boolean contains(Properties properties, String name) {		
		if (!properties.containsKey(name)) {
			logger.error("Missing property: "+name);
			return false;
		}		
		return true;
	}

	/**
	 * Writes a list of data beans into Kafka. The data MUST be in a Java beans-compliant format.
	 * Throws if list is empty.
	 */
	public final int write(List<T> data, String[] topics, long timeoutSecs) {
		if (data.isEmpty()) {
			throw new IllegalArgumentException("Empty data list.");
		}
		Objects.requireNonNull(topics, "Empty list of topics.");		
		
		int sent = 0;
		for (T t : data) {
			if (this.write(t, topics)) {
				sent++;
			}
		}

		this.waitFor(timeoutSecs);

		logger.info("Sent "+sent+" of "+data.size()+" entries to '"+this.toString(topics)+"'.");
		return sent;
	}

	private final String toString(String[] topics) {
		StringBuilder allTopics = new StringBuilder();
		for (String topic : topics) {
			allTopics.append(topic).append(", ");
		}
		return allTopics.toString();
	}
	
	private final boolean write(Object o, String[] topics) {

		for (String topic : topics) {
			try {
				byte[] payload = this.avroConverter.toAvroPayload(o);
				this.send(payload, topic);	
				logger.info("Message sent to topic: "+topic);
			}
			catch (Exception e) {
				e.printStackTrace();
				return false;
			}
		}
		return true;
	}

	private final void send(byte[] payload, String topics) throws InterruptedException, ExecutionException, TimeoutException {
		ProducerRecord<String, byte[]> record = new ProducerRecord<String, byte[]>(topics, payload);                
		this.kafkaSender.send(record);				
	}		

	private final void waitFor(long secs) {
		try {
			Thread.sleep(secs * 1000);
		} catch (InterruptedException e) {			
			e.printStackTrace();
		}
	}
}

File: src/main/scala/za/co/absa/abris/examples/deprecated/KafkaAvroWriter.scala
/*
 * Copyright 2018 ABSA Group Limited
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package za.co.absa.abris.examples.deprecated

import java.util.Properties

import org.apache.spark.sql.catalyst.encoders.RowEncoder
import org.apache.spark.sql.{Dataset, Encoder, Row}
import za.co.absa.abris.avro.format.SparkAvroConversions
import za.co.absa.abris.avro.parsing.utils.AvroSchemaUtils
import za.co.absa.abris.examples.data.generation.ComplexRecordsGenerator
import za.co.absa.abris.examples.utils.ExamplesUtils._

object KafkaAvroWriter {

  private val PARAM_JOB_NAME = "job.name"
  private val PARAM_JOB_MASTER = "job.master"
  private val PARAM_KEY_AVRO_SCHEMA = "key.avro.schema"
  private val PARAM_PAYLOAD_AVRO_SCHEMA = "payload.avro.schema"
  private val PARAM_AVRO_RECORD_NAME = "avro.record.name"
  private val PARAM_AVRO_RECORD_NAMESPACE = "avro.record.namespace"
  private val PARAM_INFER_SCHEMA = "infer.schema"
  private val PARAM_LOG_LEVEL = "log.level"
  private val PARAM_TEST_DATA_ENTRIES = "test.data.entries"
  private val PARAM_EXECUTION_REPEAT = "execution.repeat"
  private val PARAM_NUM_PARTITIONS = "num.partitions"

  def main(args: Array[String]): Unit = {

    // there is a sample properties file at /src/test/resources/DataframeWritingExample.properties
    checkArgs(args)

    val properties = loadProperties(args)

    val spark = getSparkSession(properties, PARAM_JOB_NAME, PARAM_JOB_MASTER, PARAM_LOG_LEVEL)

    import spark.implicits._

    implicit val encoder: Encoder[Row] = getEncoder

    do {
      val rows = createRows(properties.getProperty(PARAM_TEST_DATA_ENTRIES).trim().toInt)
      val dataframe = spark.sparkContext.parallelize(rows, properties.getProperty(PARAM_NUM_PARTITIONS).toInt).toDF()

      dataframe.show()

      toAvro(dataframe, properties)
        .write
        .format("kafka")
        .addOptions(properties) // 1. this method will add the properties starting with "option."; 2. security options can be set in the properties file
        .save()
    } while (properties.getProperty(PARAM_EXECUTION_REPEAT).toBoolean)
  }

  private def toAvro(dataframe: Dataset[Row], properties: Properties) = {
    import za.co.absa.abris.avro.AvroSerDe._
    if (properties.getProperty(PARAM_INFER_SCHEMA).trim().toBoolean) {
      dataframe.toAvro(properties.getProperty(PARAM_AVRO_RECORD_NAME), properties.getProperty(PARAM_AVRO_RECORD_NAMESPACE))
    } else {
      dataframe.toAvro(properties.getProperty(PARAM_PAYLOAD_AVRO_SCHEMA))
    }
  }

  private def createRows(howMany: Int): List[Row] = {
    ComplexRecordsGenerator.generateUnparsedRows(howMany)
  }

  private def getEncoder: Encoder[Row] = {
    val avroSchema = AvroSchemaUtils.parse(ComplexRecordsGenerator.usedAvroSchema)
    val sparkSchema = SparkAvroConversions.toSqlType(avroSchema)
    RowEncoder.apply(sparkSchema)
  }
}
